app/models/import.rb

```diff
@@ -1,185 +1,137 @@
+    import!
 
-  def csv
-    get_normalized_csv_with_validation
-  end
+    family.sync
 
-  def available_headers
-    get_raw_csv.table.headers
+    update! status: :complete
+  rescue => error
+    update! status: :failed, error: error.message
   end
 
-  def get_selected_header_for_field(field)
-    column_mappings&.dig(field.key) || field.key
+  def csv_rows
+    @csv_rows ||= parsed_csv
   end
 
-  def update_csv!(row_idx:, col_idx:, value:)
-    updated_csv = csv.update_cell(row_idx.to_i, col_idx.to_i, value)
-    update! normalized_csv_str: updated_csv.to_s
+  def csv_headers
+    parsed_csv.headers
   end
 
-  # Type-specific methods (potential STI inheritance in future when more import types added)
-  def publish
-    update!(status: "importing")
-
-    transaction do
-      generate_transactions.each do |txn|
-        txn.save!
-      end
-    end
-
-    self.account.sync
-
-    update!(status: "complete")
-  rescue => e
-    update!(status: "failed")
-    Rails.logger.error("Import with id #{id} failed: #{e}")
+  def csv_sample
+    @csv_sample ||= parsed_csv.first(2)
   end
 
   def dry_run
-    generate_transactions
+    {
+      transactions: rows.count,
+      accounts: Import::AccountMapping.for_import(self).creational.count,
+      categories: Import::CategoryMapping.for_import(self).creational.count,
+      tags: Import::TagMapping.for_import(self).creational.count
+    }
   end
 
-  def expected_fields
-    @expected_fields ||= create_expected_fields
+  def required_column_keys
+    []
   end
 
-  private
-
-    def get_normalized_csv_with_validation
-      return nil if normalized_csv_str.nil?
-
-      csv = Import::Csv.new(normalized_csv_str)
-
-      expected_fields.each do |field|
-        csv.define_validator(field.key, field.validator) if field.validator
-      end
-
-      csv
-    end
-
-    def get_raw_csv
-      return nil if raw_file_str.nil?
-      Import::Csv.new(raw_file_str, col_sep:)
-    end
-
-    def should_initialize_csv?
-      raw_file_str_changed? || column_mappings_changed?
-    end
+  def column_keys
+    raise NotImplementedError, "Subclass must implement column_keys"
+  end
 
-    def initialize_csv
-      generated_csv = generate_normalized_csv(raw_file_str)
-      self.normalized_csv_str = generated_csv.table.to_s
+  def generate_rows_from_csv
+    rows.destroy_all
+
+    mapped_rows = csv_rows.map do |row|
+      {
+        account: row[account_col_label].to_s,
+        date: row[date_col_label].to_s,
+        qty: row[qty_col_label].to_s,
+        ticker: row[ticker_col_label].to_s,
+        price: row[price_col_label].to_s,
+        amount: row[amount_col_label].to_s,
+        currency: (row[currency_col_label] || default_currency).to_s,
+        name: (row[name_col_label] || default_row_name).to_s,
+        category: row[category_col_label].to_s,
+        tags: row[tags_col_label].to_s,
+        entity_type: row[entity_type_col_label].to_s,
+        notes: row[notes_col_label].to_s
```

Comment:
Good question—this ensures that any "empty" cells are treated as `blank?` strings rather than `nil?`.

This helps with the `Import::Mapping` so that we can have an import mapping with `""` as the key.

Ultimately I think this could be solved in a couple different ways, but that was what I landed on.

Reply:
Clever... liked it.